### ISSUE TITLE
quiche/src/tls.rs uses an unsafe bindings strategy

### ISSUE THREAD

**Issue by davidben** [cloudflare/quiche#1559]
https://github.com/cloudflare/quiche/blob/master/quiche/src/tls.rs#L1255 contains a pile of hand-written bindings for BoringSSL functions. BoringSSL is not ABI-stable, so this is not a supported way to use the library. This will result in subtle memory errors, with no checking that you've gotten it right.

Indeed, there are already mistakes:
https://github.com/cloudflare/quiche/blob/master/quiche/src/tls.rs#L1442C8-L1443

BoringSSL uses `size_t`, not `int`, for those functions. That means this crate only works correctly on calling conventions where `int` and `size_t` are compatible. Specifically, on 64-bit platforms, you're relying on the upper half of the 64-bit register having a defined value when it is used to pass in a 32-bit parameter. You're in luck that most calling conventions require this, but this is not guaranteed.

This should be replaced with something robust, e.g. bindgen.

**Comment by ghedo** [comment]
> This should be replaced with something robust, e.g. bindgen.

Yes, though the main complication is that we don't enforce a specific version of BoringSSL and we don't necessarily have access to the headers for that version (e.g. an application can provide its own BoringSSL build). Need to figure out a way to work around that.

**Comment by davidben** [comment]
Not having headers seems unlikely to work well, given how we expect BoringSSL to be used. You also don't usually expect to be able to use a library in C without its headers.

Perhaps you all could transition to requiring the application provide both the build and the headers?

**Comment by davidben** [comment]
In parallel, I suppose I can start by upstreaming some changes to gradually take the hand-transcribed bindings closer to what bindgen would have generated. There are quite a few mistakes. :-)

**Comment by davidben** [comment]
Any news on this? This is a memory safety vulnerability waiting to happen if we, e.g., ever extend our existing structs.
